hack/build-image

```diff
@@ -0,0 +1,597 @@
+can be used to create "Fully Qualified Image Names" that combine the
+image kind (samba-server, client, etc) and a tag that indicates all
+the unique properties of the image. This includes the package source,
+the base os, and the architecture.
+
+In addition to building the images, one can push images, list images,
+and list build status files (aka buildfiles).
+
+Usage:
+  # build an image
+  ./hack/build-image --kind server --distro-base fedora --arch amd64
+
+  # print out the FQIN
+  ./hack/build-image --kind samba-server --distro-base fedora \\
+    --arch amd64 --print
+
+  # print out the FQIN and additional tags
+  ./hack/build-image --kind samba-server --distro-base fedora \\
+    --arch amd64 --print-tags
+
+  # print out the FQIN and additional tags for multiple images, with
+  # and without a repository base
+  ./hack/build-image --kind samba-server \\
+    --distro-base fedora \\
+    --distro-base centos \\
+    --distro-base opensuse \\
+    --arch amd64 \\
+    --repo-base quay.io/foobar --without-repo-bases --print-tags
+
+"""
+
+import argparse
+import logging
+import os
+import pathlib
+import platform
+import shlex
+import shutil
+import subprocess
+import sys
+
+
+logger = logging.getLogger("build-image")
+
+# IMAGE_KINDS - map aliases/names to canonical names for the kinds
+# of images we can build
+IMG_SERVER = "samba-server"
+IMG_AD_SERVER = "samba-ad-server"
+IMG_CLIENT = "samba-client"
+IMG_TOOLBOX = "samba-toolbox"
+IMAGE_KINDS = {
+    # short names
+    "server": IMG_SERVER,
+    "ad-server": IMG_AD_SERVER,
+    "client": IMG_CLIENT,
+    "toolbox": IMG_TOOLBOX,
+    # canonical names
+    "samba-server": IMG_SERVER,
+    "samba-ad-server": IMG_AD_SERVER,
+    "samba-client": IMG_CLIENT,
+    "samba-toolbox": IMG_TOOLBOX,
+}
+
+# ARCHITECTURES - map supported arch names/alias to canonical names
+AMD64 = "amd64"
+ARM64 = "arm64"
+ARCHITECTURES = {
+    # alternate names
+    "x86_64": AMD64,
+    "aarch64": ARM64,
+    # canonical names
+    "amd64": AMD64,
+    "arm64": ARM64,
+}
+
+# DISTROS - list of supported distro bases
+FEDORA = "fedora"
+CENTOS = "centos"
+OPENSUSE = "opensuse"
+DISTROS = [
+    FEDORA,
+    CENTOS,
+    OPENSUSE,
+]
+
+# PACKAGE_SOURCES - list of known package sources
+DEFAULT = "default"
+NIGHTLY = "nightly"
+PACKAGE_SOURCES = [DEFAULT, NIGHTLY]
+
+# SOURCE_DIRS - image source paths
+# (paths are relative to common image source dir)
+SOURCE_DIRS = {
+    IMG_SERVER: "server",
+    IMG_AD_SERVER: "ad-server",
+    IMG_CLIENT: "client",
+    IMG_TOOLBOX: "toolbox",
+}
+
+DEFAULT_PKG_SOURCES = [DEFAULT]
+DEFAULT_DISTRO_BASES = [FEDORA]
+LATEST = "latest"
+QUAL_NONE = "unqualified"
+QUAL_DISTRO = "distro-qualified"
+
+
+_DISCOVERED_CONTAINER_ENGINES = []
+
+
+def check_kind(kind):
+    """Return the canonical name for the image kind or raise a ValueError."""
+    try:
+        return IMAGE_KINDS[kind]
+    except KeyError:
+        raise ValueError(f"invalid kind: {kind}")
```

Comment:
Consider:
```
except KeyError as ex:
    raise ValueError(f"invalid kind: {kind}") from ex
```

Reply:
Exception chaining can be handy when you want to retain the info from the inner exception, but here, the fact that we use a dict lookup to validate the value feels like an implementation detail to me. Do you feel strongly about this?

Reply:
No; it is more a common habit I carry with me. As said in [https://legacy.python.org/dev/peps/pep-3134/](PEP-3134), "In order to debug the problem, it is useful to know about both exceptions" but I guess that here it is just a "nice to have".